### ISSUE TITLE
Inconsistent highlighting in YAML mode

### ISSUE THREAD

**Issue by krzysztof-magosa** [DarthFennec/highlight-indent-guides#17]
Hi,
For some reason plugin does not indent some levels of indentation.
It looks that when there is no character above indentation it's not treated as next level.
See screenshot.

<img width="311" alt="screen shot 2017-03-24 at 00 06 00" src="https://cloud.githubusercontent.com/assets/6112411/24273958/adb643b2-1025-11e7-9a32-fe3b030c6c39.png">

**Comment by DarthFennec** [comment]
Yes, this is by design. Indent levels are defined by the indentation of the first line in the block being more than that of the line above the block, so if lines further down are somewhere in the middle it won't consider it a different level. This works just fine in most languages, because generally the only time you'd have this is if you're wrapping a long `if` or `for` statement or something like that, in which case the extra space is technically not indentation, so it makes sense to not highlight it. I can definitely see your point about YAML though ... interesting ...

I'm seeing two possible ways to interpret this issue:

1. This could be seen as an issue with the way `highlight-indent-guides` detects indent levels, specifically that it only detects them from the top down rather than in both directions. I designed it this way on purpose: `highlight-indent-guides` is built on `font-lock`, and `font-lock` only calculates from the top down. Because `font-lock` does a really good job of deciding which parts of the buffer to recalculate and when (only calculating the visible part of the buffer, etc), I can trust it to only calculate guides when necessary. However, if I need to detect guides from the bottom up, by necessity I would need to calculate guides for the entire block from the very start. The block may be extremely large (think Java or C++ class files), so that initial calculation could take a rather long time, with very little benefit.
2. This could be seen as a lack of support for the way YAML in particular defines blocks, specifically that multiple blocks can start on the same line. On one hand, this is better: it wouldn't incur a performance hit, and also it would be more correct for this specific case (the previous solution would calculate two levels of indentation instead of one on lines 3, 4, and 5, whereas this solution would also calculate two levels on lines 19 and 20). Even so, I'm against language-specific features for this mode, so I'd only feel comfortable doing this if I found a very general and reliable way to discover the indent levels for a language at runtime.

I'm really not a fan of option 1. Even though it seems trivial and intuitive for `highlight-indent-guides` to work like this, it really fights against `font-lock`'s natural flow. Although, I notice that languages which tend to have large top-level blocks don't tend to have a need for this sort of thing, and vice versa, so I might be okay with adding this feature and recommending that people only enable it for languages that would really use it.

Option 2 seems better, but only if there's a generic way to do it. I really don't like language-specific features, because it can lead to messy code, but also because it implies that some languages are more important than others. I don't want to add a feature that one person can benefit from, if someone else can't benefit simply because they're using a different language. I think my best bet for a generic discovery system is to try and detect the tab cycles imposed by the major mode, as this should provide all reasonable indent blocks for any given situation. I can test this out at some point and see if it's actually a viable option.

I don't think I'm going to get to this any time soon, and I'm still very on the fence about whether it should be done at all. If I find something really promising I will look into it, though.

**Comment by krzysztof-magosa** [comment]
Thanks for the detailed explanation @DarthFennec.
